File: patcher.py
#!/usr/bin/env python3
import os, sys, json, shutil, datetime

# === Embedded patches.json data ===
PATCHES_JSON = {
    "verify": True,
    "patches": [
        {
            "offset": "0xB8FBF",
            "replace": "EB",
            "expect": None
        },
        {
            "offset": "0xB8FC0",
            "replace": "80",
            "expect": None
        }
    ]
}

def apply_patches(exe_path, patches):
    # Make backup folder
    backup_dir = os.path.join(os.path.dirname(exe_path), "backup")
    os.makedirs(backup_dir, exist_ok=True)

    # Copy exe to backup with timestamp
    ts = datetime.datetime.now().strftime("%Y%m%d_%H%M%S")
    backup_path = os.path.join(backup_dir, f"{os.path.basename(exe_path)}.{ts}.bak")
    shutil.copy2(exe_path, backup_path)
    print(f"[+] Backup created: {backup_path}")

    with open(exe_path, "rb+") as f:
        data = bytearray(f.read())

        for patch in patches["patches"]:
            offset = int(patch["offset"], 16)
            new_bytes = bytes.fromhex(patch["replace"])
            expect = bytes.fromhex(patch["expect"]) if patch["expect"] else None

            # Safety check
            if offset + len(new_bytes) > len(data):
                print(f"[-] Offset {hex(offset)} out of range.")
                return False

            old_bytes = data[offset:offset+len(new_bytes)]
            if expect and old_bytes != expect:
                print(f"[-] Mismatch at {hex(offset)}: expected {expect.hex()} got {old_bytes.hex()}")
                return False

            print(f"[+] Patching {hex(offset)}: {old_bytes.hex()} -> {new_bytes.hex()}")
            data[offset:offset+len(new_bytes)] = new_bytes

        # Write patched file
        f.seek(0)
        f.write(data)
    print("[+] Patching complete.")
    return True

def main():
    exe_path = "Ripcord.exe"
    if not os.path.exists(exe_path):
        print("[-] Ripcord.exe not found in this directory.")
        sys.exit(1)

    success = apply_patches(exe_path, PATCHES_JSON)
    sys.exit(0 if success else 1)

if __name__ == "__main__":
    main()

File: patcher.c
#include <stdio.h>
#include <stdlib.h>
#include <direct.h> // For _mkdir on Windows

// Patch offsets and bytes 
const long OFFSETS[] = {0xB8FBF, 0xB8FC0};
const unsigned char PATCH_BYTES[] = {0xEB, 0x80};
const int PATCH_COUNT = 2;

int main() {
    // Create backup directory
    if (_mkdir("backup") != 0) {
        printf("Note: Backup directory already exists or could not be created.\n");
    }

    // Verify Ripcord.exe exists
    FILE *src = fopen("Ripcord.exe", "rb");
    if (!src) {
        printf("Error: Could not find Ripcord.exe. Ensure it's in the same folder.\n");
        return 1;
    }

    // Verify file is large enough for patching
    fseek(src, 0, SEEK_END);
    long file_size = ftell(src);
    if (file_size < OFFSETS[PATCH_COUNT - 1] + 1) {
        printf("Error: Ripcord.exe is too small or invalid.\n");
        fclose(src);
        return 1;
    }
    rewind(src);

    // Backup Ripcord.exe
    FILE *backup = fopen("backup\\Ripcord.exe", "wb");
    if (!backup) {
        printf("Error: Could not create backup file.\n");
        fclose(src);
        return 1;
    }

    // Copy file to backup
    char buffer[4096];
    size_t bytes;
    while ((bytes = fread(buffer, 1, sizeof(buffer), src)) > 0) {
        if (fwrite(buffer, 1, bytes, backup) != bytes) {
            printf("Error: Failed to write backup file.\n");
            fclose(src);
            fclose(backup);
            return 1;
        }
    }
    fclose(src);
    fclose(backup);
    printf("Backup created at backup\\Ripcord.exe\n");

    // Patch Ripcord.exe
    FILE *file = fopen("Ripcord.exe", "r+b");
    if (!file) {
        printf("Error: Could not open Ripcord.exe for patching.\n");
        return 1;
    }

    // Apply patches
    for (int i = 0; i < PATCH_COUNT; i++) {
        if (fseek(file, OFFSETS[i], SEEK_SET) != 0) {
            printf("Error: Could not seek to offset 0x%lx.\n", OFFSETS[i]);
            fclose(file);
            return 1;
        }
        if (fwrite(&PATCH_BYTES[i], 1, 1, file) != 1) {
            printf("Error: Could not write byte 0x%02x at offset 0x%lx.\n", PATCH_BYTES[i], OFFSETS[i]);
            fclose(file);
            return 1;
        }
        printf("Patched byte 0x%02x at offset 0x%lx\n", PATCH_BYTES[i], OFFSETS[i]);
    }

    fclose(file);
    printf("Patching complete. You can now use Ripcord.exe.\n");
    return 0;

}
